File: lib/server/ai-client.js
const DEFAULT_BASE_URL = "https://api.deepseek.com/v1";

export function normalizeBaseUrl(baseUrl) {
  return (baseUrl || DEFAULT_BASE_URL).replace(/\/+$/, "");
}

export function safeJsonParse(value) {
  try {
    return JSON.parse(value);
  } catch {
    return null;
  }
}

export function extractDeltaText(payload) {
  const choice = payload?.choices?.[0];
  const delta = choice?.delta?.content ?? choice?.message?.content ?? "";

  if (typeof delta === "string") {
    return delta;
  }

  if (Array.isArray(delta)) {
    return delta
      .map((part) => {
        if (typeof part === "string") {
          return part;
        }
        if (part?.type === "text" && typeof part.text === "string") {
          return part.text;
        }
        return "";
      })
      .join("");
  }

  return "";
}

export function extractMessageText(payload) {
  const content = payload?.choices?.[0]?.message?.content ?? "";

  if (typeof content === "string") {
    return content;
  }

  if (Array.isArray(content)) {
    return content
      .map((part) => {
        if (typeof part === "string") {
          return part;
        }
        if (part?.type === "text" && typeof part.text === "string") {
          return part.text;
        }
        return "";
      })
      .join("");
  }

  return "";
}

export function createTextStream(upstreamBody) {
  const decoder = new TextDecoder();
  const encoder = new TextEncoder();

  return new ReadableStream({
    async start(controller) {
      const reader = upstreamBody.getReader();
      let buffer = "";

      const pushEventBlock = (eventBlock) => {
        const lines = eventBlock.split(/\r?\n/);

        for (const line of lines) {
          if (!line.startsWith("data:")) {
            continue;
          }

          const data = line.slice(5).trim();
          if (!data) {
            continue;
          }

          if (data === "[DONE]") {
            return true;
          }

          const parsed = safeJsonParse(data);
          if (!parsed) {
            console.error("无法解析上游流式数据:", data);
            continue;
          }

          const text = extractDeltaText(parsed);
          if (text) {
            controller.enqueue(encoder.encode(text));
          }

          if (parsed?.choices?.[0]?.finish_reason) {
            return true;
          }
        }

        return false;
      };

      try {
        while (true) {
          const { done, value } = await reader.read();
          if (done) {
            break;
          }

          buffer += decoder.decode(value, { stream: true });

          while (true) {
            const matched = buffer.match(/\r?\n\r?\n/);
            if (!matched || matched.index === undefined) {
              break;
            }

            const boundary = matched.index;
            const eventBlock = buffer.slice(0, boundary);
            buffer = buffer.slice(boundary + matched[0].length);

            if (pushEventBlock(eventBlock)) {
              controller.close();
              return;
            }
          }
        }

        if (buffer.trim()) {
          pushEventBlock(buffer);
        }

        controller.close();
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function getTextModel() {
  return process.env.AI_MODEL || "deepseek-chat";
}

export function getVisionModel() {
  return process.env.AI_VISION_MODEL || "";
}

export async function requestAiChatCompletion({
  messages,
  stream = false,
  temperature = 0.7,
  model,
  maxTokens,
}) {
  const apiKey = process.env.AI_API_KEY;

  if (!apiKey) {
    const error = new Error("API Key 未配置");
    error.code = "missing_api_key";
    throw error;
  }

  const response = await fetch(`${normalizeBaseUrl(process.env.AI_BASE_URL)}/chat/completions`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: model || getTextModel(),
      stream,
      temperature,
      max_tokens: maxTokens,
      messages,
    }),
  });

  return response;
}

export function extractJsonObject(text) {
  if (typeof text !== "string") {
    return null;
  }

  const fencedMatch = text.match(/```json\s*([\s\S]*?)```/i);
  if (fencedMatch?.[1]) {
    const parsed = safeJsonParse(fencedMatch[1].trim());
    if (parsed) {
      return parsed;
    }
  }

  const start = text.indexOf("{");
  const end = text.lastIndexOf("}");
  if (start === -1 || end === -1 || end <= start) {
    return null;
  }

  return safeJsonParse(text.slice(start, end + 1));
}

File: backend/app/services/ai_client.py
from __future__ import annotations

from typing import Any

import httpx

from app.config import Settings


class AIClient:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings
        self.base_url = settings.ai_base_url.rstrip("/")
        self.headers = {
            "Authorization": f"Bearer {settings.ai_api_key}",
            "Content-Type": "application/json",
        }

    async def list_models(self) -> list[str]:
        async with httpx.AsyncClient(timeout=self.settings.ai_timeout) as client:
            response = await client.get(f"{self.base_url}/models", headers=self.headers)
            response.raise_for_status()
            payload = response.json()
        models = payload.get("data", [])
        return [item["id"] for item in models if item.get("supported_endpoint_types")]

    async def resolve_model(self) -> str:
        if self.settings.ai_model:
            return self.settings.ai_model

        models = await self.list_models()
        preferred_order = [
            "gpt-4.1-mini",
            "gpt-5-mini",
            "gpt-4o-mini",
            "gpt-4.1",
        ]
        for model in preferred_order:
            if model in models:
                return model
        if not models:
            raise RuntimeError("当前 AI 接口未返回可用模型，请检查接口地址或密钥配置。")
        return models[0]

    async def complete(
        self,
        messages: list[dict[str, str]],
        *,
        temperature: float | None = None,
        max_tokens: int = 4000,
    ) -> tuple[str, str]:
        model = await self.resolve_model()
        payload: dict[str, Any] = {
            "model": model,
            "messages": messages,
            "temperature": self.settings.ai_temperature if temperature is None else temperature,
            "max_tokens": max_tokens,
        }

        async with httpx.AsyncClient(timeout=self.settings.ai_timeout) as client:
            response = await client.post(
                f"{self.base_url}/chat/completions",
                headers=self.headers,
                json=payload,
            )
            response.raise_for_status()
            data = response.json()

        message = data["choices"][0]["message"]["content"]
        if isinstance(message, list):
            text = "".join(
                part.get("text", "")
                for part in message
                if isinstance(part, dict)
            )
        else:
            text = str(message)
        return model, text.strip()
